File: Java/src/main/java/com/toaction/hot100/linkedList/LRUCache.java
package com.toaction.hot100.linkedList;

import java.util.HashMap;
import java.util.Map;

public class LRUCache {

    /*
    * https://leetcode.cn/problems/lru-cache/
    * */


    /**
     双向链表存放具体值，Map快速定位元素
     */
    class ListNode{
        int key;
        int val;
        ListNode next;
        ListNode pre;
        public ListNode(){};
        public ListNode(int key, int val){
            this.key = key;
            this.val = val;
            this.next = null;
            this.pre = null;
        }
    }

    private int capacity;
    private int size;
    private Map<Integer, ListNode> cache = new HashMap<>();
    private ListNode head;
    private ListNode tail;

    public LRUCache(int capacity) {
        this.capacity = capacity;
        this.size = 0;
        this.head = new ListNode(0, 0);
        this.tail = new ListNode(0, 0);
        this.head.next = this.tail;
        this.tail.pre = this.head;
    }

    public int get(int key) {
        ListNode node = cache.getOrDefault(key, null);
        if(node == null){
            return -1;
        }
        // 移动到链表头部
        moveTohead(node);
        return node.val;
    }

    public void put(int key, int value) {
        // key已存在，变更数据值
        ListNode node = cache.getOrDefault(key, null);
        if(node != null){
            node.val = value;
            moveTohead(node);
            return;
        }
        // 已满，删除一个节点
        if(size == capacity){
            ListNode temp = tail.pre;
            deleteNode(temp);
            cache.remove(temp.key);
            size--;
        }
        // 插入新节点
        node = new ListNode(key, value);
        cache.put(key, node);
        insertHead(node);
        size++;
    }

    // 先删除节点，然后插入到头部
    private void moveTohead(ListNode node){
        deleteNode(node);
        insertHead(node);
    }

    private void deleteNode(ListNode node){
        ListNode temp = node.pre;
        temp.next = node.next;
        node.next.pre = temp;
    }

    private void insertHead(ListNode node){
        ListNode temp = head.next;
        head.next = node;
        node.next = temp;
        temp.pre = node;
        node.pre = head;
    }
}

File: linkedlist/lru_cache.go
package linkedlist

/*
146. LRU 缓存：https://leetcode.cn/problems/lru-cache/

请你设计并实现一个满足  LRU (最近最少使用) 缓存 约束的数据结构。
实现 LRUCache 类：
LRUCache(int capacity) 以 正整数 作为容量 capacity 初始化 LRU 缓存
int get(int key) 如果关键字 key 存在于缓存中，则返回关键字的值，否则返回 -1 。
void put(int key, int value) 如果关键字 key 已经存在，则变更其数据值 value ；
如果不存在，则向缓存中插入该组 key-value 。
如果插入操作导致关键字数量超过 capacity ，则应该 逐出 最久未使用的关键字。
*/

type DListNode struct {
	key   int
	value int
	prev  *DListNode
	next  *DListNode
}

type LRUCache struct {
	capacity int
	size     int
	data     map[int]*DListNode
	head     *DListNode
	tail     *DListNode
}

func Constructor(capacity int) LRUCache {
	cache := LRUCache{
		capacity: capacity,
		size:     0,
		data:     make(map[int]*DListNode),
		head:     &DListNode{},
		tail:     &DListNode{},
	}
	cache.head.next = cache.tail
	cache.tail.prev = cache.head
	return cache
}

// 若存在，将节点移动到 head
func (cache *LRUCache) Get(key int) int {
	DListNode := cache.data[key]
	if DListNode == nil {
		return -1
	}
	// 删除节点
	cache.remove(DListNode)
	// 插入到头部
	cache.addHead(DListNode)
	return DListNode.value
}

// 从链表中删除节点
func (cache *LRUCache) remove(DListNode *DListNode) {
	DListNode.prev.next, DListNode.next.prev = DListNode.next, DListNode.prev
	cache.size--
}

// 将节点添加到链表首部
func (cache *LRUCache) addHead(DListNode *DListNode) {
	DListNode.prev = cache.head
	DListNode.next = cache.head.next
	cache.head.next.prev = DListNode
	cache.head.next = DListNode
	cache.size++
}

func (cache *LRUCache) Put(key int, value int) {
	if v, ok := cache.data[key]; ok {
		v.value = value
		cache.Get(key)
		return
	}
	// 删除尾部节点
	if cache.size == cache.capacity {
		DListNode := cache.tail.prev
		delete(cache.data, DListNode.key)
		cache.remove(DListNode)
	}
	// 插入
	DListNode := &DListNode{key: key, value: value}
	cache.data[key] = DListNode
	cache.addHead(DListNode)
}
